src/network/bufferedconnection.h

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (C) 2008-2019 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#ifndef WL_NETWORK_BUFFEREDCONNECTION_H
+#define WL_NETWORK_BUFFEREDCONNECTION_H
+
+#include <memory>
+#include <mutex>
+
+#include "network/network.h"
+#include "network/relay_protocol.h"
+
+/**
+ * A wrapper around a network connection.
+ * Does not contain logic but provides a buffer to read and write
+ * RelayCommand / uint8_t / std::string / SendPacket from/to the network stream.
+ *
+ * Offers the Peeker class to check whether the required data has already been received
+ * before trying to read the data.
+ */
+class BufferedConnection {
+public:
+	/**
+	 * Allows to check whether the required data is completely in the buffer before starting to
+	 * receive it.
+	 *
+	 * The idea of this methods is that the caller can check whether a complete command
+	 * can be received before starting to remove data from the buffer.
+	 * Otherwise, the caller would have to maintain their own buffer.
+	 *
+	 * The methods will not remove any bytes, but will check whether the required value can be read.
+	 * After checking, an internal peek-pointer will be incremented. The next call of a method will
+	 * then start reading after the previous value. Only a successful peek will move the pointer.
+	 *
+	 * Note that a successful peek does not mean that the received bytes really are of the requested
+	 * type, it only means they could be interpreted that way. Whether the type matches is in the
+	 * responsibility of the caller.
+	 *
+	 * Using any method of this class will trigger an internal read from the network socket inside
+	 * the given BufferedConnection.
+	 *
+	 * \warning Calling any receive() method on the given connection will invalidate the Peeker.
```

Comment:
So there should be some `bool isValid()` Method?

Reply:
I could add one, but it would probably relatively complicated.
Currently the Peeker is only used for at most a few consecutively lines, so I would keep the warning as a "callers responsibility". Adding a isValid() method doesn't make it any better anyway I guess, since the caller would have to remember to call that method.